Detect offline clusters

weyfonk opened the pull request:
This adds a cluster status monitor to the Fleet controller, which checks when each cluster last saw its agent online. If more than the expected interval elapses, that cluster is considered offline, and the monitor updates its bundle deployments' statuses to reflect that. This will trigger status updates to bundles, `GitRepo`s, clusters and cluster groups.

Refers to #594.

Open points:
* how far, and how fine-grained, do we want to make bundle deployment status updates for offline clusters? This currently takes a fairly basic approach, updating both `Ready` and `Monitored` conditions while clearing modified and non-ready statuses, to prevent outdated messages from appearing in a bundle deployment's display status and further up the chain of status updates (to bundles, then upwards to `GitRepo`s, etc)
* should we make the frequency of monitoring configurable?
* do we have a way to exclude the local/management cluster from agent-last-seen checks?


### Update 2026-03
* the cluster controller checks for offline bundle deployments, to mark the corresponding cluster as offline. It also supports the opposite, namely marking a cluster back online when its bundle deployments are detected as such.
* this is covered by integration tests, not involving the cluster monitor, while end-to-end tests validate this behaviour when the cluster monitor is involved.

weyfonk committed:
Add cluster status monitor to controller

This allows the Fleet controller to detect offline clusters and update
statuses of bundle deployments targeting offline clusters.

Next to do:
* understand how bundle deployment status updates should be propagated
  to bundles (which currently simply appear as `Modified`) and further
  up
* write tests (eg. integration tests, updating cluster status by hand?)
* set sensible defaults (eg. monitoring interval higher, and threshold
  higher than agent's cluster status reporting interval)

weyfonk committed:
Reflect offline cluster state in more bundle deployment status fields

This enables that state to be reflected upwards in bundle, GitRepo,
cluster and cluster group statuses.

weyfonk committed:
Move cluster status monitor to separate package

This also provides unit tests for detecting offline clusters.

weyfonk committed:
Set sensible configuration for cluster status monitor

The frequency of cluster status checks is currently hard-coded to 1
minute, but could be made configurable.

The threshold for considering a cluster offline now explicitly depends
on how often agents report their statuses to the management cluster.
Changes to that configured interval should impact the cluster status
monitor, which would take the new value into account from its next run
onwards.

weyfonk committed:
Eliminate linting errors

This fixes an error and ignores a few others to make the linter happy.

0xavi0 commented on internal/cmd/controller/clustermonitor/monitor.go:
```diff
@@ -0,0 +1,132 @@
+					case "Ready":
+						// FIXME: avoid relying on agent pkg for this?
+						mc := monitor.Cond(v1alpha1.BundleDeploymentConditionReady)
+						mc.SetError(&bd.Status, "Cluster offline", fmt.Errorf("cluster is offline"))
+						// XXX: do we want to set Deployed and Installed conditions as well?
+						// XXX: should we set conditions to `Unknown`?
+					case "Monitored":
+						mc := monitor.Cond(v1alpha1.BundleDeploymentConditionMonitored)
+						mc.SetError(&bd.Status, "Cluster offline", fmt.Errorf("cluster is offline"))
+
+					}
+				}
+
+				return c.Status().Update(ctx, t)
```
Shouldn't it be setting the Error on `t` instead of `bd`?
The update call is done with `t`.

weyfonk replied:
Nice catch, thanks! This still behaved as expected because of references, but was clearly not ideal. Fixed.

0xavi0 commented on internal/cmd/controller/clustermonitor/monitor_test.go:
```diff
@@ -0,0 +1,369 @@
+	for _, cond := range bdStatus.Conditions {
+		if cond.Type == "Ready" {
+			foundReady = true
+
+			if !strings.Contains(cond.Message, "offline") {
+				return false
+			}
+		} else if cond.Type == "Monitored" {
+			foundMonitored = true
+
+			if !strings.Contains(cond.Message, "offline") {
+				return false
+			}
```
Would be good to check that the condition's status is `False` when the `offline` string is found and `True` otherwise. 
Also would be good to check that the Reason in the Condition.

weyfonk replied:
Thanks, checks added.

0xavi0 reviewed (commented):
I don't know if you want to merge this PR with the comments with open questions, maybe better to clarify first. 

Just a couple of observations, but overall LGTM

weyfonk committed:
Check for condition status and reason

This adds checks ensuring that for offline clusters, for which calls to
update bundle deployment statuses are expected, those statuses contain
`Ready` and `Monitored` conditions with status `False` and reasons
reflecting the cluster's offline status.

weyfonk committed:
Prevent agent check-in interval from being 0

This ensures that creating a new agent bundle fails with an agent
check-in interval set to 0.

weyfonk committed:
Prevent check-in interval from being 0 when importing cluster

This adds a check on the agent check-in interval to cluster import, for
consistency with agent bundle updates.

weyfonk committed:
Make cluster monitor interval and threshold configurable

This enables users to determine how often the Fleet controller will
check for offline clusters, and based on which threshold. If the
configured threshold is below the triple of the check-in interval, that
tripled value will be used instead.

weyfonk committed:
Skip bundle deployment updates for already offline clusters

This optimises updates to bundle deployments, running them only against
clusters which bundle deployments are not yet marked as offline.

weyfonk committed:
Set Ready condition to Unknown for offline clusters

This better reflects what is then known about workloads running in such
clusters than `False`.

weyfonk committed:
Fix json attribute for cluster monitor interval

This should fix Fleet controller deployments complaining about the
interval being 0 when it should never be.

weyfonk committed:
Run cluster status monitor on unsharded controller only

Running one cluster status monitor per Fleet controller pod is not
necessary and may cause conflicts in sharded setups.

weyfonk committed:
Add agent check-in interval to agent install tests

Omitting the agent check-in interval when patching the
`fleet-controller` config map would now lead to errors when setting up
agents with a check-in interval bearing the default value for a
duration, ie 0s.
That interval is now set with a hard-coded value, which is of no
importance for such tests, for the sake of not being zero.

weyfonk committed:
Fix linting errors

weyfonk committed:
Use feature flag for enabling cluster monitor

The cluster monitor is disabled by default, but may be enabled by
setting Helm value `clusterMonitor.enabled` to `true`.

weyfonk committed:
Add threshold to cluster status check log message

This may help troubleshoot unexpected threshold values, for instance
caused by high agent check-in intervals.

weyfonk committed:
Mark Cluster as Offline when its bundle deployments are offline

While the cluster monitor only updates bundle deployment statuses,
the cluster reconciler handles updates to cluster statuses, to prevent
concurrent status updates.

weyfonk committed:
Prevent cluster updates triggered by cluster status updates

Similarly to how a bundle should not be reconciled based on its own
status updates, neither should a cluster.

weyfonk committed:
Use patch operations in cluster monitor

Updates to bundle deployment statuses should be more lightweight and
less invasive.

This commit also adds WIP end-to-end tests to validate bundle deployment
status updates for offline clusters, which the cluster monitor handles,
followed by cluster status updates, triggered by the cluster reconciler.

weyfonk committed:
Validate offline cluster status changes in both directions

New end-to-end tests cover a cluster becoming offline, checking that the
cluster itself and its bundle deployments are recognised as such, brings
the cluster back online, and runs those checks again.

Actual clusters coming back online may not translate in `fleet-agent`
deployments being scaled back up, as is the case in this simulation.

weyfonk committed:
Detect online clusters from online bundle deployments

The cluster controller can now detect that a previously offline cluster
has online bundle deployments, and marks the cluster as online
accordingly.

weyfonk committed:
Make linters happy

Following a rebase against the latest state of `main`, a few errors had
popped up.

weyfonk committed:
Disconnect cluster instead of scaling Fleet agent deployment

Taking advantage of k3d, multi-cluster end-to-end tests are able to
simulate disconnection of a cluster by disconnecting the cluster node
container from its Docker network, and reconnecting it.

Disconnecting the server node is enough, as this blocks outbound
traffic, in this case the agent's check-in requests to the controller.
Disconnecting the load balancer node, to block inbound traffic, is not
necessary.

This should represent a more realistic use case than scaling Fleet agent
deployments.

weyfonk committed:
Address Copilot's comments

weyfonk committed:
Make linters happy

A couple of comments had been made obsolete by the previous commit.

0xavi0 commented on internal/cmd/controller/clustermonitor/monitor.go:
```diff
@@ -0,0 +1,156 @@
+					fallthrough
+				case "Monitored":
+					if cond.Message == v1alpha1.ClusterOfflineMsg {
+						break bd_update
```
I'm not sure I understand this.
Isn't this break statement also breaking the outer loop?

weyfonk replied:
Yes, I seem to recall the idea was that if _one_ of the bundle deployments was already marked offline, then we could assume that _all_ bundle deployments for that cluster had been marked offline by a previous iteration, in which case we could break out of the outer loop, as we wouldn't need to update the rest of bundle deployments.

Come to think of it, there's still a risk that one or more new bundle deployment(s) might have been created since the previous run of this loop, hence the above assumption might not hold.
Therefore, I've changed this `break` statement to a `continue`, which simply skips to the next bundle deployment.

0xavi0 commented on internal/cmd/controller/reconciler/cluster_controller.go:
```diff
@@ -222,6 +245,28 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 			helmOps[helmOpKey] = (state == fleet.Ready) || helmOps[helmOpKey]
 		}
+
+		if summary.MessageFromCondition(fleet.ClusterConditionReady, bd.Status.Conditions) == fleet.ClusterOfflineMsg {
+			// bundle deployment offline, mark cluster offline
+			if isClusterOffline {
+				continue
+			}
+
+			isClusterOffline = true
+
+			for _, c := range offlineConds {
+				cond := monitor.Cond(c)
+				cond.SetError(&cluster.Status, "Cluster offline", errors.New(fleet.ClusterOfflineMsg))
+				cond.Unknown(&cluster.Status)
+			}
+
+		} else if summary.MessageFromCondition(fleet.ClusterConditionReady, cluster.Status.Conditions) == fleet.ClusterOfflineMsg {
+			// bundle deployment online, mark offline cluster back online
+			for _, c := range offlineConds {
+				cond := monitor.Cond(c)
+				cond.SetError(&cluster.Status, "", nil)
+			}
+		}
 	}
```
nit: If there are 2 Bundle deployments and they are in the slice in the following order:

bdA (offline)
bdC (not offline)

bdA will be processed and will call `cond.SetError(&cluster.Status, "Cluster offline", errors.New(fleet.ClusterOfflineMsg))`

bdB will be then processed and will cleanup the previous status, right?


I understand this should be temporary and at some point both BDs will report as "cluster offline" if the cluster is really offline, but I guess that it's maybe better to set the status to "offline" if any of the BDs is offline.

weyfonk replied:
Good point; I've simplified this to mark the cluster offline if _any_ bundle deployment is offline, and to only mark it back online if _all_ BDs are back online. This should prevent flapping cluster status updates.

0xavi0 reviewed (commented):
lgtm overall, just a couple of questions.

weyfonk committed:
Skip single bundle deployment if already offline instead of whole set

When a bundle deployment is already marked offline, updating it again to
mark it offline would be wasteful. However, other bundle deployments may
still require that update, especially those which may have been created
since the offline bundle deployment update logic was last run.

weyfonk committed:
Simplify setting cluster offline and back online from BDs

A cluster should be marked offline if _any_ of its bundle deployments is
offline, and back online when all of them are back online. This prevents
flapping status updates in case not all bundle deployments have been
updated by the cluster monitor.

weyfonk commented:
Closing this for now, as the current [approach](https://github.com/rancher/fleet/blob/main/internal/cmd/controller/summary/summary.go#L91-L95) of reporting deployed, but obsolete bundle deployments as `WaitApplied` instead of `ErrApplied` is considered more efficient than an additional offline cluster detector.
